### DISCUSSION TITLE
What is the appetite for breaking backwards-compatibility with .NET Framework 1.x? (yes, the 20-year-old one)

### DISCUSSION THREAD

**Discussion by daiplusplus** [dotnet/csharplang#4293] [Language Ideas] [answered]
Now that the .NET Framework 1.0 (and 1.1?) is old enough to drink in the US, I'd like to raise the perennial question of "can we please eliminate pre-generic types that are difficult to deal with - to make the rest of our lives easier?"

I'm not a fan of Apple's "backwards compatibility? Never heard of it!"-attitude towards legacy software, but I do like their willingness to break with their past mistakes when it helps their platform (namely: Swift) - in that spirit I'd like to nudge the .NET BCL design team in the direction of **removing** legacy _non-generic_ interfaces from .NET 6 (or at least marking them as `[Obsolete]` in .NET 6 before nuking them in .NET 7).

Here's a non-exhaustive list:

* `IComparable`
* `IList`
* `IConvertible`
* `IEnumerable`

My motivation concerns algebraic types, actually.

I've been working on an implementation of static algebraic data-types built on-top of Roslyn's new Code Generation facilities (the capabilities of the platform are amazing, so kudus to everyone involved!), but I noticed something rather... disturbing and/or annoying:

As a test-case, I have a Discriminated Union type of `Int32 | String` - and my ADT implementation detects that both `Int32` and `String` implement `IComparable`, so it provides for _forwarding_ the implementation, however the legacy `IComparable` interfaces are insufficiently expressive enough to be useful _at all_ for use in ADTs because one has to consult their _human-readable documentation_ to see that:

> The parameter, `obj`, must be the same type as the class or value type that implements this interface; otherwise, an `ArgumentException` is thrown.

...which makes a mockery of ADTs in a nominative type system.

Below is pseudocode of what my Roslyn code-generator generates for `Int32 | String`'s implementation of `IComparable`:

```
private IComparable AsIComparable()
{
    switch( this.index )
    {
    case 1: return this.int32;
    case 2: return this.stringValue;
    default: throw new InvalidOperationException( "Invalid state." );
    }
}

IComparable.Compare( Object obj )
{
    IComparable self = this.AsIComparable();
    return self.Compare( obj );
}
```

Which means that this use-case fails when it's _reasonable_ to assume it should not:

```
( Int32 | String ) oneOfInt32OrString_but_is_actually_Int32 = 123;
( Int32 | String ) oneOfInt32OrString_but_is_actually_String = "abc";

Int32 cmp = oneOfInt32OrString_but_is_actually_Int32.CompareTo( oneOfInt32OrString_but_is_actually_String  );
```

So I'm faced with a decision that I'd rather I didn't have to make at all:

* Either imbue my ADT `Union` type with _secret knowledge_ of these legacy interfaces such that it **doesn't** reimplement them.
* Or imbue my ADT `Union` type with _secret knowledge_ of how to _correctly_ implement these legacy interfaces to include type-checking so that they can avoid a nasty runtime `ArgumentException`.

I note the generic `IComparable<T>` solves this problem because the `T` type-parameter is sufficiently expressive enough for code-generation to *know* that incompatible types cannot be compared.

------

The bottom-line:

Besides vague hand-waving excuses of _backwards-compatibility_, is there any *substantiated* and *reasonable* justification to maintain these legacy interfaces when looking at .NET 6 and .NET 7?

As an op-ed: I don't want C# to end-up like Java where crucial and drastic improvements in developer productivity and program-correctness are sidelined for 15+ years because of appeals to _backwards compatibility_ that are completely bunk when you consider that literally no-one is running unmodified bytecode last compiled in 2002 on a modern JVM today, and no-doubt the thousands of security changes to the JVM since then will have broken enough things by now. Given that moving from the .NET Framework to .NET Core was a massive breaking-change _anyway_ we should stop pretending that .NET's livelihood depends on _Pursuit Bank_ and _Province Farm Insurance_ being able to retarget and recompile their codebases from source without making any changes at all.

**Comment by YairHalberstadt** [comment, marked as answer]
This is an issue for dotnet/runtime not csharplang

However I really don't think this is possible - whilst annoying, removing these interfaces would break 99% of all existing .Net applications.

For example, every existing implementation of `IEnumerable<T>` also has to implement `IEnumerable.GetEnumerator()`. Removing the `IEnumerable` interface will mean every single implementation of `IEnumerable<T>` will now fail to compile.

**Reply by daiplusplus** [reply]
> Removing the `IEnumerable` interface will mean every single implementation of `IEnumerable<T>` will now fail to compile.

That's why I proposed marking it as `[Obsolete]` in .NET 6 before removing it in .NET 7.

And the fix for a failed compile is to just delete the 3 lines of code that forward `IEnumerable.GetEnumerator()` to `IEnumerable<T>.GetEnumerator()`. I don't believe that's a _significant_ breaking-change to warrant keeping a detrimental interface around on life-support.

Besides, now that the CLR supports _default interface implementation_, as a stopgap we could let `IEnumerable<T>` provide the base stub implementation of `IEnumerable` before eventually removing it.

**Reply by YairHalberstadt** [reply]
> That's why I proposed marking it as [Obsolete] in .NET 6 before removing it in .NET 7.

That doesn't really help. Upgrades need to be painless in order to encourage users to upgrade. Making every single application change just isn't an option.

Also that's just one example. `Enumerable.Cast` only requires one type parameter because it makes use of the `IEnumerable` interface. Removing this interface would require changing every usage to specify two type parameters.

https://source.dot.net/#System.Linq/System/Linq/Cast.cs

**Reply by daiplusplus** [reply]
> > That's why I proposed marking it as [Obsolete] in .NET 6 before removing it in .NET 7.
> 
> That doesn't really help. Upgrades need to be painless in order to encourage users to upgrade. Making every single application change just isn't an option.

It is keeping with SemVer though.

> Also that's just one example. `Enumerable.Cast` only requires one type parameter because it makes use of the `IEnumerable` interface. Removing this interface would require changing every usage to specify two type parameters.
> 
> https://source.dot.net/#System.Linq/System/Linq/Cast.cs

That argument is making an excuse for C#'s lack of support for partial generic type parameter inference - if C# did support partial generic parameter arguments then it wouldn't be an issue (and FWIW, there does exist a workaround by abusing type-inference for an `out` parameter:

```
public static IEnumerable<TOut> Cast<TIn,TOut>( this IEnumerable<TIn> source, out TOut _ )
{
    // ...
}

// Usage:
IEnumerable<Object> input = ...
IEnumerable<String> output = input.Cast( out String _ ); // Explicit generic type arguments are inferred from `input` and `out TOut`.
```

**Reply by HaloFour** [reply]
> It is keeping with SemVer though.

The runtime does not adopt SemVer, nor should it.

Breaking every single application ever written (and every single dependency of every single application) is simply an untenable ask.

**Reply by daiplusplus** [reply]
> The runtime does not adopt SemVer, nor should it.

Right - sorry, I wasn't suggesting that .NET Core / .NET's versioning should be based on SemVer, but what I was saying was that surely making _some_ breaking changes is appropriate in new _major releases_ (of course with an appropriate notice-period, official deprecation, etc).

> Breaking every single application ever written (and every single dependency of every single application) is simply an untenable ask.

Given that .NET Core is not backwards-compatible with the .NET Framework I feel you're misrepresenting my position.  I am not suggesting that anything will suddenly break from installing a routine software update: I'm saying there should be a deprecation policy for the parts of .NET that have outlived their usefulness and are fast becoming a maintenance burden.

> Breaking every single application ever written

_That's hyperbole_. Modern applications written since .NET Framework 2.0 introduced the generic interfaces in 2005 won't be using the legacy non-generic collections interfaces as part of any core or important functionality - if they're even using those interfaces at all.

And this discussion could have been avoided entirely if .NET Core 1.0 moved the older interfaces to a type-forwarded 
shim/compatibility assembly and tagged them with the `[Obsolete]` attribute.

**Reply by HaloFour** [reply]
> Given that .NET Core is not backwards-compatible with the .NET Framework I feel you're misrepresenting my position.

Most of what was removed being entire subsystems of functionality that were both no longer useful and would be burdensome to rewrite.  They didn't introduce breaks for the sake of introducing breaks or to such fundamental parts of functionality.  Every break fractures the ecosystem, and .NET Core, despite the efforts to bridge the compatibility gaps, has yet to catch up to .NET Framework when it comes to active development.

> Modern applications written since .NET Framework 2.0 introduced the generic interfaces in 2005 won't be using the legacy non-generic collections interfaces as part of any core or important functionality - if they're even using those interfaces at all.

Yet WPF and ASP.NET Core both use and expose APIs based on those non-generic interfaces.  I'd suggest that perhaps they have not outlived their usefulness.

**Reply by daiplusplus** [reply]
> Most of what was removed being entire subsystems of functionality that were both no longer useful and would be burdensome to rewrite. They didn't introduce breaks for the sake of introducing breaks or to such fundamental parts of functionality. Every break fractures the ecosystem, and .NET Core, despite the efforts to bridge the compatibility gaps, has yet to catch up to .NET Framework when it comes to active development.

"has yet to catch up to .NET Framework when it comes to active development." - I assume by this you're referring to the ISV ecosystem, not development of .NET itself? 

> > Modern applications written since .NET Framework 2.0 introduced the generic interfaces in 2005 won't be using the legacy non-generic collections interfaces as part of any core or important functionality - if they're even using those interfaces at all.
> 
> Yet WPF and ASP.NET Core both use and expose APIs based on those non-generic interfaces. I'd suggest that perhaps they have not outlived their usefulness.

ASP.NET Core doesn't expose `IEnumerable` though. I just loaded all of the ASP.NET Core 2.2 assemblies on my computer into ILSpy and did a Used-By/Exposed-By Analysis run on `IEnumerable` and while ASP.NET Core does *use* the legacy `IEnumerable` in _some_ places (mostly from usage of Linq's `OfType` and `Cast` extension-methods which _require_ `IEnumerable` and are undefined for `IEnumerable<T>`), it doesn't expose any public APIs _at all_ that use `IEnumerable` , screenshot proof:

![image](https://user-images.githubusercontent.com/1693078/103555253-d27c7500-4e64-11eb-925d-c1ed2bddcc4e.png)

**Reply by YairHalberstadt** [reply]
As explained, every single application that defines an implementation of `IEnumerable<T>` (i.e. 99% of all applications) will break on upgrading. No matter how easy it is to fix that break, that simply isn't an option.

And that's just one of the breaks. Every application that uses `Enumerable.Cast` or `OfType` (again 99% of applications) will be broken in a different way, with a different (even if it is easy) fix.

Yes, of course if we were doing this from scratch we would do things differently. But to justify a break as big as this (note that this will break more applications than the move from .Net Framework to .Net Core) the benefits would have to be absolutely overwhelming. The argument that you would have to embed your SourceGenerator with "secret knowledge of these legacy interfaces" doesn't come to a millionth of the amount of justification that a move such as this would need.

Not that they had the opportunity to make breaks such as this when they started .Net Core, but except for a few exceptions, didn't, because the benefit of making it easy for people to switch/upgrade by far outweighs the benefits of removing legacy interfaces, no matter how annoying.

**Reply by daiplusplus** [reply]
> As explained, every single application that defines an implementation of `IEnumerable<T>` (i.e. 99% of all applications) will break on upgrading. No matter how easy it is to fix that break, that simply isn't an option.

That assumes the CLR would never be adapted to gracefully handle that case - and do you have an opinion on my proposal that the BCL add `[Obsolete]` to the older interfaces and/or moving them to a shim assembly?

**Reply by YairHalberstadt** [reply]
Adding Obsolete doesn't help much: For those that care about warnings, they still have to fix everything on moving to .Net 6, for those that don't they have to fix everything on the move to .Net 7. It also ignores that many people jump versions, straight from .NET 5 to .NET 7 for example. Obsolete can be a useful mitigation when deprecating APIs, but it is only a mitigation and just isn't enough to change the equation here.

> That assumes the CLR would never be adapted to gracefully handle that case

It's not just the CLR, it's the compiler as well that will need to accept code implementing `IEnumerable` when `IEnumerable` doesn't exist. I'm not sure what kind of graceful handling would be possible here. If the CLR, and compiler were somehow able to make sure that all existing code that used any of these legacy interfaces continued to work then this would certainly be much more doable, but I imagine that's impossible without just keeping these interfaces.

**Comment by YairHalberstadt** [comment]
Note that for your particular use case you will probably want to special case `IComparable<T>` anyway, as I think it would be reasonable to compare an `int32` with an `(int32 | string)`. As such I don't think special casing `IComparable` is particularly crazy.

**Reply by daiplusplus** [reply]
My ADT implementation does not generate an `IComparable<T>` implementation because `Int32` and `String` don't share an `IComparable<T>` interface implementation.

**Comment by canton7** [comment]
I'll just leave these here -- usages of those interfaces within the runtime:

- [`IEnumerable`](https://source.dot.net/#System.Private.CoreLib/F/workspace/_work/1/s/bin/repo/runtime/src/libraries/System.Private.CoreLib/src/System/Collections/IEnumerable.cs/IEnumerable.cs,9be451ac13d86a97,references)
- [`IList`](https://source.dot.net/#System.Private.CoreLib/F/workspace/_work/1/s/bin/repo/runtime/src/libraries/System.Private.CoreLib/src/System/Collections/IList.cs/IList.cs,5d74f6adfeaf6c7d,references)
- [`IComparable`](https://source.dot.net/#System.Private.CoreLib/IComparable.cs,29a39323fbfa896b,references)

`IEquatable` doesn't exist. There is no generic version of `IConvertible`.

**Reply by GalaxiaGuy** [reply]
`IEnumerable` is also quite fundamental to Xamarin Forms and WPF.

**Reply by daiplusplus** [reply]
> `IEnumerable` is also quite fundamental to Xamarin Forms and WPF.

`IEnumerable` is equivalent to `IEnumerable<Object>` though, and because `IEnumerable<out T>` is covariant that means it's legal (and _always correct_) to cast from any `IReadOnlyList<T>` or `IList<T>` and `IEnumerable<T>` to `IEnumerable<Object>` because `Object` is the root type for all types that the old `IEnumerable` can accommodate anyway (and `IEnumerable<T>` has the advantage of not boxing value-types if you use `where T : struct` unlike with `IEnumerable` which *always* boxes).

WinForms, WebForms, WPF, and others all share the `public Object DataSource { get; set; }` property idiom which I never liked because you had to refer to the documentation see what *actual* types were supported - which was not consistent between implementations (e.g. some would use reflection to look for `Count` and `Item[]` indexers, or `GetEnumerator` (with or without nominative `IEnumerable` implementation, etc).

My argument is that wherever `System.Collections.IEnumerable` appears, it *should* be a straightforward fix to replace it with `System.Collections.Generic.IEnumerable<Object>` - and then replace `Object` with an appropriate `T` if time allows.

**Reply by HaloFour** [reply]
Covariance does not apply to value types, `IEnumerable<int>` cannot be cast to `IEnumerable<object>` or vice versa.

**Reply by daiplusplus** [reply]
> Covariance does not apply to value types, `IEnumerable<int>` cannot be cast to `IEnumerable<object>` or vice versa.

Indeed - but the rule against covariance with value vs. reference-types is [couched in identity-preservation](https://stackoverflow.com/questions/12454794/why-covariance-and-contravariance-do-not-support-value-type) which is an implementation detail.

A workaround is straightforward: allow for implicit conversions between interface types that don't preserve identity (of both the object implementing the interface, and of the individual boxed values - given that value-types have to be boxed with the old `IEnumerable` I don't imagine this will affect the correctness of software that currently accepts an `IEnumerable` to receive boxed value-types.

For example (and supposing this was valid...):

```
public interface IEnumerable<out T>
{
    static implicit operator IEnumerable<Object>( IEnumerable<T> self ) where T : struct
    {
        return new BoxingEnumerableAdapter<T>( self );
    }
}

class BoxingEnumerableAdapter<T> : IEnumerable<Object> {

    public IEnumerator<Object> GetEnumerator() => // etc
}

// ditto for converting from IEnumerator<T> to IEnumerator<Object>
```

**Reply by HaloFour** [reply]
The better answer is in Eric's own response to a similar question:  https://stackoverflow.com/questions/4096299/why-doesnt-delegate-contravariance-work-with-value-types

> The whole point of having _covariant reference conversions_ is to _preserve reference identity_.

Which would indicate that it's not an implementation detail, it's the desired behavior.

Having that cast instead interpreted as a custom conversion would also break the rule that no custom conversion operation should be able to redefine existing conversions.

**Reply by daiplusplus** [reply]
> I'll just leave these here -- usages of those interfaces within the runtime:
> 
> * [`IEnumerable`](https://source.dot.net/#System.Private.CoreLib/F/workspace/_work/1/s/bin/repo/runtime/src/libraries/System.Private.CoreLib/src/System/Collections/IEnumerable.cs/IEnumerable.cs,9be451ac13d86a97,references)
> * [`IList`](https://source.dot.net/#System.Private.CoreLib/F/workspace/_work/1/s/bin/repo/runtime/src/libraries/System.Private.CoreLib/src/System/Collections/IList.cs/IList.cs,5d74f6adfeaf6c7d,references)
> * [`IComparable`](https://source.dot.net/#System.Private.CoreLib/IComparable.cs,29a39323fbfa896b,references)
> 
> `IEquatable` doesn't exist. There is no generic version of `IConvertible`.

I went through the uses of `IEnumerable` and `IList` that you linked to. While there is no generic version of `IConvertible`, I still feel it's a poorly-designed legacy interface with little future applicability: while `IConvertible` certainly makes _some_ numeric conversions possible, it's nonsensical for `System.String` to implement `IConvertible` when you cannot meaningfully convert a `String` using `IConverible.ToDouble`, for example.

* The majority of references to `IEnumerable` are in code that needs to implement `IEnumerable.GetEnumerator()` because they implement `IEnumerable<T>`.
* Many of the implementations of `IList` have accompanying comments that admit it should be using `IList<T>` but can't for other reasons, such as:
    > `InputGestureCollection` - Collection of `InputGestures`. Stores the InputGestures sequentially in a genric InputGesture List Will be changed to generic List implementation once the parser supports generic collections.
    * ...and most of the other references are for libraries that were imported from .NET Framework with minimal modification or those that need it _because other types need it_ because they haven't been updated yet either, such as `DesignerOptionCollection`.

I'm not saying the legacy interfaces aren't being used _at all_ - I'm saying we should add the `[Obsolete]` attribute to them and devise a plan to get rid of them _eventually_.

**Comment by CyrusNajmabadi** [comment]
This sounds like a *lot* of pain.  And that part has been discussed in depth.  What i'm not seeing is an explanation/discussion of hte benefits that warrant this pain.

**Reply by daiplusplus** [reply]
> What i'm not seeing is an explanation/discussion of hte benefits that warrant this pain.

Succinctly:

* [The C# team is aware of strong interest to add support for](https://github.com/dotnet/csharplang/discussions/317) first-class algebraic data-types, higher-kinded polymorphism and other features to the type-system and I think it's safe to say that C# will end-up getting those features within a decade given how other languages (Rust, TypeScript, Swift, etc) are also adopting them.
* But I observed that using ADTs (such as a Union/Or-type) with types in the BCL today that use these legacy interfaces causes problems because these legacy interfaces seem to violate _Liskov's Substitution Principle_
    *  e.g. a method that accepts an `Int32` via its `IComparable` interface cannot then compare it with a `String` via its own `IComparable` interface because the two implementations are incompatible.
* So given that I've experienced a _gotcha_ with these legacy interfaces when I thought I was doing something ordinarily benign - and given that this violation of the _Principle of Least Astonishment_ will doubtlessly impact a significant number of users in the event that ADTs do become a first-class feature of C#/.NET I think we should plan for that now.

So the benefit is that when C# _does_ eventually have built-in first-class support for a statement like below, there won't be any nasty surprises or unfortunate workarounds:

```
( Int32 | String ) isInt = 123;
( Int32 | String ) isStr = "abc";
IComparable a = isInt;
IComparable b = isStr;
Console.WriteLine( a.CompareTo( b ) ); // Throws ArgumentException
```

By removing `IComparable` and other interfaces that the compiler-alone cannot determine should be type-compatible would help, I feel - though I imagine there are more reasonable alternatives - I swear I'm not obsessed with wanting to raze the BCL every few years... even if it sure does look like it.

**Reply by YairHalberstadt** [reply]
Wouldn't a less burn the world to the ground solution simply be to change the BCL implementations of IComparable to return false rather than throw when the types don't match?

Such a change is likely to break very few people.

**Reply by HaloFour** [reply]
What does this have to do with ADTs?  The following has been legal (and throws) since C# 1.0:

```cs
Int32 isInt = 123;
String isStr = "abc";
IComparable a = isInt;
IComparable b = isStr;
Console.WriteLine( a.CompareTo( b ) ); // Throws ArgumentException
```

That also assumes how union types might be implemented by the compiler or runtime, either of which could take that interface into consideration.

**Reply by daiplusplus** [reply]
> Wouldn't a less burn the world to the ground solution simply be to change the BCL implementations of IComparable to return false rather than throw when the types don't match?
> 
> Such a change is likely to break very few people.

Arguably that would also break software too - in more subtle and less-visible ways which some people would argue is worse than a very visible build-breakage though.

**Reply by YairHalberstadt** [reply]
> Arguably that would also break software too - in more subtle and less-visible ways which some people would argue is worse than a very visible build-breakage though.

Certainly there might be some software which expects an ArgumentException to be thrown, catches it, and handles it, as part of ordinary execution flow. Such would break every principle of good code development but it probably exists. However I estimate that's an absolutely tiny percentage of all code, as opposed to almost every existing project.

Whether it's worth it I don't know, but it's certainly at least within the realms of possibility, unlike the current proposal.

**Reply by daiplusplus** [reply]
> What does this have to do with ADTs? The following has been legal (and throws) since C# 1.0:

Sorry I should have given a better example.

The documentation for `IComparable` allows for implementations to throw when faced with an incompatible comparison argument - I believe it's then implied that provided the immediate type of the implementation of `IComparable` is the same then the `CompareTo` method should not throw an exception.

The problem is that with a union-type it means the exception is harder to predict in code, for example:

```
( Int32 | String ) a = GetInt32OrStringAtRandom();
( Int32 | String ) b = GetInt32OrStringAtRandom();

// Programmer's internal monologue: "both 'a' and 'b' have the same declared type: `( Int32 | String )`, and `IComparable` should not throw when `this` object and `CompareTo` arguments have the same type, therefore `a.CompareTo( b )` should not throw.

Console.WriteLine( a.CompareTo( b ) ); 

// Programmer's internal monologue: "why did the above code just throw? both 'a' and 'b' have the same type!"
```

> That also assumes how union types might be implemented by the compiler or runtime, either of which could take that interface into consideration.

I raised that in my opening post: while having the compiler or runtime be specifically _aware_ of those older interfaces, it doesn't work as a general-solution.

**Reply by HaloFour** [reply]
Considering that there is no championed proposal to explore union types that would look or behave as you've described I don't see why the exercise to prune the runtime at this point would make any sense.  The only form of union types being explored are discriminated unions, and if whatever the compiler emits for those implements `IComparable` (which I doubt) then the compiler can handle the implementation appropriately at that time.

**Reply by canton7** [reply]
> Wouldn't a less burn the world to the ground solution simply be to change the BCL implementations of IComparable to return false rather than throw when the types don't match?

It returns an int, not a bool. All possible ints are currently valid return values.
